pkg/restore/conflicts.go

```diff
@@ -0,0 +1,50 @@
+/*
+Copyright 2018 the Heptio Ark contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package restore
+
+import (
+	"github.com/heptio/ark/pkg/util/collections"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"strings"
+)
+
+// mergeSecrets takes image pull secrets and secrets from a backed up service accounts and merges them into the current in-cluster service account
+func mergeSecrets(fromCluster, fromBackup *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	secrets, _ := collections.GetSlice(fromBackup.UnstructuredContent(), "secrets")
+	for i := len(secrets) - 1; i >= 0; i-- {
+		secret := secrets[i]
+		val, _ := collections.GetString(secret.(map[string]interface{}), "name")
+		if strings.HasPrefix(val, "default-token-") {
+			secrets = append(secrets[:i], secrets[i+1:]...)
+		}
+	}
+	fromBackup.Object["secrets"] = secrets
+	areEqual := equality.Semantic.DeepEqual(fromCluster, fromBackup)
+
+	var desired *unstructured.Unstructured
+	if !areEqual {
+		desired = fromCluster.DeepCopy()
+		desired.Object["imagePullSecrets"] = fromBackup.Object["imagePullSecrets"]
```

Comment:
is there any reason we'd want to merge these lists? IOW, is there any scenario where `fromCluster` already has `imagePullSecrets` that we'd want to keep? Or, is this function really purpose-built for the scenario where `fromCluster` has exactly one `secret` (the default SA token) and zero `imagePullSecrets`?  I'd just think about and capture the assumptions, either in comments or code.

Reply:
We should probably merge the image pull secrets. My assumption was that the serviceaccount created by Kubernetes wouldn't have any image pull secretss associated by default, though.